### DISCUSSION TITLE
Node support for apt

### DISCUSSION THREAD

**Discussion by TracerDS** [nodejs/help#4622] [Q&A] [answered]
Hi, I wanted to download node for my ubuntu server, but currently only external package managers (`nvm`, `fnm`) and prebuilt binaries are supported.
Would it be possible to make a `.deb` package for debian based distros for download?
From that you could just use `apt install ./node.deb`. No external package managers.

**Comment by avivkeller** [comment]
Node.js doesn't offer an official apt package for installation. However, external organizations like NodeSource provide such packages.

There's also an apt package named `nodejs` available at [Debian's package tracker](https://tracker.debian.org/pkg/nodejs), but it is not an officially supported/maintained.

For official installation methods, visit the [Node.js download page](https://nodejs.org/en/download).

---

I'm not sure who to ping to see if this will ever be added in the future, but I've asked on the OpenJSF slack.

I've also marked this as the answer, but if new information that outdates this, I'll undo that.

**Comment by aduh95** [comment]
We do list nvm and fnm even though it's not an official installation method and is not maintained by us. So it might make sense to list an apt option as well. That can be discussed in nodejs/nodejs.org repo.

Regarding creating a `.deb`, I'm not aware of any plan for doing this; it could happen if someone champions the initiative and is ready to convince the project it's worth investigating resources for (we host files for all the releases of Node.js, which adds up to storage costs that we don't have unlimited budget for), but that seems like a hard sell tbh.

**Comment by mhdawson** [comment]
I also see this as outside the scope of the project.

There are many different linux distros and I think it's up to those ecosystems to create/package Node.js (for example this is an rpm for Fedora/RHEL etc.) In the past other groups (like NodeSource) had also created packages outside the scope of the project.  So there may be a way to get Node.js through apt but not from the binaries produced by the project.

It's also worth noting that you can manuallyl install the linux binaries provided by the project as they should run most distros including recent ubuntu versions.

**Reply by TracerDS** [reply]
But third party managers that you install from curl (like nvm) dont install node as they should. Instead they just link node to the userdir that installed node.
If you do it as root then all the binaries will sit in /root/.local directory.

Creating an apt package would ease this process and will actually install everything correctly.
I tried doing it manually via prebuilt binaries but they dont match the content from apt repo (missing files, libraries, etc.).

**Reply by mhdawson** [reply]
I don't disagree that having an apt package will make it easier, but I don't think the Node.js project can take on the scope for building/managing one for a linux distros. One approach might be to see if you can get the debian/ubuntu community to take that on?

**Reply by TracerDS** [reply]
> I don't disagree that having an apt package will make it easier, but I don't think the Node.js project can take on the scope for building/managing one for a linux distros. One approach might be to see if you can get the debian/ubuntu community to take that on?

If it wouldnt be possible to have nodejs publish an apt package, maybe there could be steps to build one to install it yourself?
That way .deb package would still be available, harder to get but available nonetheless

**Reply by mhdawson** [reply]
@TracerDS if you wanted to create those instructions (or find somebody who will) and the propose linking to those instructions somewhere in the website or project that might work.

**Comment by mateuspaulino** [comment, marked as answer]
Hi @TracerDS, it is possible but it will take a few steps:

```
sudo apt update
sudo apt install -y build-essential fakeroot dpkg-dev
```


Download Node.js Source:
Get the source code for the version of Node.js you want to package.

```
VERSION=v16.16.0  # Replace with your desired version
wget https://nodejs.org/dist/$VERSION/node-$VERSION.tar.gz
tar -xzf node-$VERSION.tar.gz
cd node-$VERSION
```
Build Node.js from source.

```
./configure
make -j$(nproc)
```
Install Node.js into a directory which will be used to create the .deb package.

```
mkdir -p ../nodejs-staging/usr/local
make install DESTDIR=$(pwd)/../nodejs-staging/usr/local
```
Create the necessary files for the .deb package.

```
    mkdir -p ../nodejs-staging/DEBIAN

cat <<EOF > ../nodejs-staging/DEBIAN/control
Package: nodejs
Version: ${VERSION:1}
Section: web
Priority: optional
Architecture: amd64
Depends: libc6 (>= 2.17), libstdc++6 (>= 5.2)
Maintainer: Your Name youremail@example.com
Description: Node.js - JavaScript runtime built on Chrome's V8 JavaScript engine
Node.js is a platform built on Chrome's JavaScript runtime for easily building fast, scalable network applications.
EOF
```


6. **Build the .deb Package**:
Use `dpkg-deb` to build the package.

```
cd ../
dpkg-deb --build nodejs-staging
```

    Now you can install the .deb package using apt.

`    sudo apt install ./nodejs-staging.deb`

**Reply by TracerDS** [reply]
Can these steps still be added to [nodejs.org website](https://nodejs.org/en/download) despite [#6502](https://github.com/nodejs/nodejs.org/pull/6502)?

**Reply by ovflowd** [reply]
It can. Just not on the "Package Manager" page now. We are in the process of rewriting the Package Managers main Download page (the one with the fancy code editors)

To properly separate OS package managers (such as brew, apt, yum) and the Node version managers (nvm, fnm, etc.)

Hence, we are not accepting new changes at the moment. But we have a dedicated page for **all** package managers for now.

**Reply by TracerDS** [reply]
> It can. Just not on the "Package Manager" page now. We are in the process of rewriting the Package Managers main Download page (the one with the fancy code editors)
> 
> To properly separate OS package managers (such as brew, apt, yum) and the Node version managers (nvm, fnm, etc.)
> 
> Hence, we are not accepting new changes at the moment. But we have a dedicated page for **all** package managers for now.

Is the page available for everyone or just maintainers?

**Reply by ovflowd** [reply]
https://nodejs.org/en/download/package-manager/all it is referred on the bottom of the Package Manager pages
